internal/database/typeddocument_test.go

```diff
@@ -109,3 +112,144 @@ func TestTypedDocumentUnmarshal(t *testing.T) {
+			wantErr: assert.NoError,
+		},
+		{
+			name: "clear hcp",
+			args: args{
+				typedDoc: &typedDocument{
+					ResourceType: api.ClusterResourceType.String(),
+				},
+				// cannot unmarshal string into Go struct field HCPOpenShiftClusterProperties.hcpOpenShiftCluster.properties.version of type api.VersionProfile
+				innerDoc: &ResourceDocument{
+					CustomerDesiredState: map[string]any{
+						"alligator": "adept",
+						"hcpOpenShiftCluster": map[string]any{
+							"bat": "black",
+							"properties": map[string]any{
+								"version": map[string]any{
+									"id": "1.2.3",
+								},
+							},
+						},
+					},
+					ServiceProviderState: map[string]any{
+						"alligator": "adept",
+						"hcpOpenShiftCluster": map[string]any{
+							"bat": "black",
+							"properties": map[string]any{
+								"dns": map[string]any{
+									"baseDomain": "example.com",
+								},
+							},
+						},
+					},
+				},
+				documentFilter: FilterHCPClusterState,
+			},
+			// verifying that we filter out all unknown fields and serialize equivalently
+			want: jsonFor(
+				&typedDocument{
+					ResourceType: api.ClusterResourceType.String(),
+				},
+				&ResourceDocument{},
+				CustomerDesiredHCPClusterState{
+					HCPOpenShiftCluster: api.HCPOpenShiftCluster{
+						Properties: api.HCPOpenShiftClusterProperties{
+							Version: api.VersionProfile{
+								ID: "1.2.3",
+							},
+						},
+					},
+				},
+				ServiceProviderHCPClusterState{
+					HCPOpenShiftCluster: api.HCPOpenShiftCluster{
+						Properties: api.HCPOpenShiftClusterProperties{
+							DNS: api.DNSProfile{
+								BaseDomain: "example.com",
+							},
+						},
+					},
+				},
+			),
+			wantErr: assert.NoError,
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			got, err := resourceDocumentMarshal(tt.args.typedDoc, tt.args.innerDoc, tt.args.documentFilter)
+			if !tt.wantErr(t, err, fmt.Sprintf("resourceDocumentMarshal(%v, %v, %v)", tt.args.typedDoc, tt.args.innerDoc, tt.args.documentFilter)) {
+				return
+			}
+			t.Log(string(got))
+			assert.Equalf(t, tt.want, string(got), "resourceDocumentMarshal(%v, %v, %v)", tt.args.typedDoc, tt.args.innerDoc, tt.args.documentFilter)
+		})
+	}
+}
+
+func jsonFor(typedDocument *typedDocument, doc *ResourceDocument, customer, serviceProvider any) string {
+	doc = toResourceDocument(doc, customer, serviceProvider)
+	return string(must(typedDocumentMarshal(typedDocument, doc)))
+}
+
+func toResourceDocument(doc *ResourceDocument, customer, serviceProvider any) *ResourceDocument {
+	doc.CustomerDesiredState = must(objToMap(customer))
+	doc.ServiceProviderState = must(objToMap(serviceProvider))
+	return doc
+}
+
+// must is a helper function that takes a value and error, returns the value if no error occurred,
+// or panics if an error occurred. This is useful for test setup where we don't expect errors.
+func must[T any](v T, err error) T {
+	if err != nil {
+		panic(err)
+	}
+	return v
+}
```

Comment:
Minor, but it jumped out at me.  Could write this instead as:
```suggestion
func must[T any](t testing.T, v T, err error) T {
    require.NoError(t, err)
	return v
}
```
where `require` is "[github.com/stretchr/testify/require](https://pkg.go.dev/github.com/stretchr/testify/require)".

I thought of writing a helper function like this but couldn't think of a good name for it.  "Must" is perfect.  I might move this to `/internal/api/testhelpers.go` so it's more widely available.